lib/srv/desktop/rdp/rdpclient/client.go

```diff
@@ -0,0 +1,345 @@
+		return trace.BadParameter("missing Addr in rdpclient.Options")
+	}
+	if o.InputMessage == nil {
+		return trace.BadParameter("missing InputMessage in rdpclient.Options")
+	}
+	if o.OutputMessage == nil {
+		return trace.BadParameter("missing OutputMessage in rdpclient.Options")
+	}
+	return nil
+}
+
+// Client is the RDP client.
+type Client struct {
+	opts Options
+
+	clientWidth, clientHeight uint16
+	username                  string
+	rustClientRef             *C.Client
+	wg                        sync.WaitGroup
+	closeOnce                 sync.Once
+}
+
+// New creates and connects a new Client based on opts.
+func New(opts Options) (*Client, error) {
+	c := &Client{opts: opts}
+
+	if err := c.readClientUsername(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if err := c.readClientSize(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if err := c.connect(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	c.start()
+	return c, nil
+}
+
+func (c *Client) readClientUsername() error {
+	for {
+		msg, err := c.opts.InputMessage()
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		u, ok := msg.(deskproto.ClientUsername)
+		if !ok {
+			logrus.Debugf("Expected ClientUsername message, got %T", msg)
+			continue
+		}
+		c.username = u.Username
+		return nil
+	}
+}
+
+func (c *Client) readClientSize() error {
+	for {
+		msg, err := c.opts.InputMessage()
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		s, ok := msg.(deskproto.ClientScreenSpec)
+		if !ok {
+			logrus.Debugf("Expected ClientScreenSpec message, got %T", msg)
+			continue
+		}
+		c.clientWidth = uint16(s.Width)
+		c.clientHeight = uint16(s.Height)
+		return nil
+	}
+}
+
+func (c *Client) connect() error {
+	addr := C.CString(c.opts.Addr)
+	defer C.free(unsafe.Pointer(addr))
+	username := C.CString(c.username)
+	defer C.free(unsafe.Pointer(username))
+
+	// *Temporary* hack for injecting passwords until we implement cert-based
+	// authentication.
+	// TODO(awly): remove this after certificates are implemented.
+	passwordStr := os.Getenv("TELEPORT_DEV_RDP_PASSWORD")
+	if passwordStr == "" {
+		return trace.BadParameter("missing TELEPORT_DEV_RDP_PASSWORD env var and certificate authentication is not implemented yet")
+	}
+	password := C.CString(passwordStr)
+	defer C.free(unsafe.Pointer(password))
+
+	res := C.connect_rdp(
+		addr,
+		username,
+		password,
+		C.uint16_t(c.clientWidth),
+		C.uint16_t(c.clientHeight),
+	)
+	if err := cgoError(res.err); err != nil {
+		return trace.Wrap(err)
+	}
+	c.rustClientRef = res.client
+	return nil
+}
+
+func (c *Client) start() {
+	// Video output streaming worker goroutine.
+	c.wg.Add(1)
+	go func() {
+		defer c.wg.Done()
+		defer c.closeConn()
+		defer logrus.Info("RDP output streaming finished")
+
+		clientRef := registerClient(c)
+		defer unregisterClient(clientRef)
+
+		if err := cgoError(C.read_rdp_output(c.rustClientRef, C.int64_t(clientRef))); err != nil {
+			logrus.Warningf("Failed reading RDP output frame: %v", err)
```

Comment:
Nit: I'd consider making logger a part of the `Client` with proper component set and do `c.log` everywhere.